src/tlo/methods/collect_event_chains.py

```diff
@@ -0,0 +1,370 @@
+        self.row_before = pd.Series()
+        self.mni_instances_before = False
+        self.mni_row_before = {}
+        self.entire_mni_before = {}
+        
+    PARAMETERS = {
+        # Options within module
+        "generate_event_chains": Parameter(
+            Types.BOOL, "Whether or not we want to collect chains of events for individuals"
+        ),
+        "modules_of_interest": Parameter(
+            Types.LIST, "Restrict the events collected to specific modules. If *, print for all modules"
+        ),
+        "events_to_ignore": Parameter(
+            Types.LIST, "Events to be ignored when collecting chains"
+        ),
+        }
+        
+    def initialise_simulation(self, sim):
+        notifier.add_listener("simulation.pop_has_been_initialised", self.on_notification_pop_has_been_initialised)
+        notifier.add_listener("simulation.on_birth", self.on_notification_of_birth)
+        notifier.add_listener("event.about_to_run", self.on_notification_event_about_to_run)
+        notifier.add_listener("event.has_just_ran", self.on_notification_event_has_just_ran)
+        
+    def read_parameters(self, resourcefilepath: Optional[Path] = None):
+        self.load_parameters_from_dataframe(pd.read_csv(resourcefilepath/"ResourceFile_GenerateEventChains/parameter_values.csv"))
+        
+    def initialise_population(self, population):
+        # Use parameter file values by default, if not overwritten
+        self.generate_event_chains = self.parameters['generate_event_chains'] \
+            if self.generate_event_chains is None \
+            else self.generate_event_chains
+            
+        self.modules_of_interest = self.parameters['modules_of_interest'] \
+            if self.modules_of_interest is None \
+            else self.modules_of_interest
+            
+        self.events_to_ignore = self.parameters['events_to_ignore'] \
+            if self.events_to_ignore is None \
+            else self.events_to_ignore
+            
+        # If modules of interest is '*', set by default to all modules included in the simulation
+        if self.modules_of_interest == ['*']:
+            self.modules_of_interest = list(self.sim.modules.keys())
+
+    def on_birth(self, mother, child):
+        # Could the notification of birth simply take place here?
+        pass
+        
+    def on_notification_pop_has_been_initialised(self, data):
+
+        # When logging events for each individual to reconstruct chains,
+        # only the changes in individual properties will be logged.
+        # At the start of the simulation + when a new individual is born,
+        # we therefore want to store all of their properties
+        # at the start.
+        if self.generate_event_chains:
+
+            # EDNAV structure to capture status of individuals at the start of the simulation
+            ednav = df_to_EAV(self.sim.population.props, self.sim.date, 'StartOfSimulation')
+
+            logger.info(key='event_chains',
```

Comment:
Looking at the output, it's not in the right format. Currently, each cell of the logged entry is a dict. 

What we want is a flat table of entries with columns: E, A, V, EventName. And one single entity+attribute+value+EventName per row. So, I think you need to loop over each of the changed attributes and log them one at a time. Otherwise, you'll have the same problem of unknown varying number of items in each logged row.